File: src/SchedulingSolver.cpp
/*
 * Copyright 2020 Maximilian Azendorf
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "SchedulingSolver.h"

#include <memory>
#include <utility>

#include "Rng.h"
#include "Util.h"
#include "Options.h"
#include "Scheduling.h"

int SchedulingSolver::calculate_available_max_push(vector<int> const& choiceScramble, int depth)
{
    int push = 0;
    for(; depth < choiceScramble.size(); depth++)
    {
        push += _inputData->choice(choiceScramble[depth]).max;
    }

    return push;
}

bool SchedulingSolver::satisfies_critical_sets(map<int, int> const& decisions, vector<CriticalSet> const& criticalSets)
{
    set<int> coveredSlots;
    for(CriticalSet const& set : criticalSets)
    {
        coveredSlots.clear();
        int missing = 0;

        for(int element : set.elements())
        {
            auto slotIt = decisions.find(element);

            if(slotIt == decisions.end())
            {
                missing++;
            }
            else if (slotIt->second != Scheduling::NOT_SCHEDULED)
            {
                coveredSlots.insert(slotIt->second);
            }
        }

        if(coveredSlots.size() + missing < _inputData->slot_count())
        {
            return false;
        }
    }

    return true;
}

bool SchedulingSolver::satisfies_scheduling_constraints(int choice, int slot, map<int, int> const& decisions)
{
    for(Constraint constraint : _inputData->scheduling_constraints(choice))
    {
        switch(constraint.type())
        {
            case ChoiceIsInSlot: if(slot != constraint.right()) return false; break;
            case ChoiceIsNotInSlot: if(slot == constraint.right()) return false; break;

            case ChoicesAreInSameSlot:
            {
                int other = constraint.left() == choice ? constraint.right() : constraint.left();
                auto otherSlotIt = decisions.find(other);
                if(otherSlotIt == decisions.end()) break;

                if(otherSlotIt->second != slot)
                {
                    return false;
                }
                break;
            }

            case ChoicesAreNotInSameSlot:
            {
                int other = constraint.left() == choice ? constraint.right() : constraint.left();
                auto otherSlotIt = decisions.find(other);
                if (otherSlotIt == decisions.end()) break;

                if(otherSlotIt->second == slot && slot != Scheduling::NOT_SCHEDULED)
                {
                    return false;
                }
                break;
            }

            case ChoicesHaveOffset:
            {
                int other = constraint.left() == choice ? constraint.right() : constraint.left();
                int offset = other == constraint.left() ? -constraint.extra() : constraint.extra();
                auto otherSlotIt = decisions.find(other);
                if(otherSlotIt == decisions.end()) break;

                if((otherSlotIt->second == Scheduling::NOT_SCHEDULED) != (slot == Scheduling::NOT_SCHEDULED))
                {
                    // Choices that have an offset must either both be scheduled or both not be scheduled.
                    return false;
                }
                if(slot == Scheduling::NOT_SCHEDULED)
                {
                    // The offset does not matter if the choice is not scheduled.
                    break;
                }
                if(otherSlotIt->second - slot != offset)
                {
                    return false;
                }

                int minSlot = std::max(0, 0 - offset);
                int maxSlot = std::min(_inputData->slot_count(), _inputData->slot_count() - offset);

                if(slot < minSlot || slot >= maxSlot) return false;
                break;
            }

            case SlotHasLimitedSize:
            {
                if(constraint.left() != slot) continue;

                // These will only be checked at the end of the backtracking (for now).
                // Todo: Maybe implement earlier constraint infeasibility detection Gt and Geq.
                //
                if(constraint.extra() == Neq || constraint.extra() == Gt || constraint.extra() == Geq) break;

                int limit = constraint.right() - (constraint.extra() == Lt ? 1 : 0);

                for(auto const& decision : decisions)
                {
                    if(decision.second == slot) limit--;
                }

                if(limit < 1) return false;

                break;
            }

            default: throw std::logic_error("Unknown scheduling type " + str(constraint.type()) + ".");
        }
    }

    if(decisions.size() + 1 == _inputData->choice_count())
    {
        // This is the last decision to be made, time to check slot size constraints.
        //
        return check_slot_size_constraints(choice, slot, decisions);
    }
    else
    {
        return true;
    }
}

bool SchedulingSolver::check_slot_size_constraints([[maybe_unused]] int choice, int slot, map<int, int> const& decisions)
{
    if (slot < 0) return true;

    // these will be counted lazily so we don't have to compute slot sizes if there are no slot size constraints.
    //
    vector<int> slotSizes;

    for(Constraint constraint : _inputData->scheduling_constraints())
    {
        if(constraint.type() != SlotHasLimitedSize) continue;

        if(slotSizes.empty())
        {
            slotSizes.resize(_inputData->slot_count());
            slotSizes[slot]++;
            for(int w = 0; w < _inputData->choice_count(); w++)
            {
                if(choice == w) continue;
                int wslot = decisions.at(w);
                if (wslot != Scheduling::NOT_SCHEDULED)
                {
                    slotSizes[wslot]++;
                }
            }
        }

        bool valid = false;
        switch(constraint.extra())
        {
            case Eq: valid = slotSizes[constraint.left()] == constraint.right(); break;
            case Neq: valid = slotSizes[constraint.left()] != constraint.right(); break;
            case Lt: valid = slotSizes[constraint.left()] < constraint.right(); break;
            case Leq: valid = slotSizes[constraint.left()] <= constraint.right(); break;
            case Gt: valid = slotSizes[constraint.left()] > constraint.right(); break;
            case Geq: valid = slotSizes[constraint.left()] >= constraint.right(); break;
        }

        if(!valid) return false;
    }

    return true;
}

bool SchedulingSolver::has_impossibilities(map<int, int> const& decisions, int availableMaxPush)
{
    for(int s = 0; s < _inputData->slot_count(); s++)
    {
        int sum = availableMaxPush;
        for(auto const& decision : decisions)
        {
            if(decision.second != s) continue;
            sum += _inputData->choice(decision.first).max;
        }

        if(sum < _inputData->chooser_count())
        {
            return true;
        }
    }

    return false;
}

vector<int>
SchedulingSolver::calculate_critical_slots(map<int, int> const& decisions, int availableMaxPush, int choice)
{
    vector<int> criticalSets;

    for(int s = 0; s < _inputData->slot_count(); s++)
    {
        int sum = availableMaxPush - _inputData->choice(choice).max;
        for(auto const& decision : decisions)
        {
            if(decision.second != s) continue;
            sum += _inputData->choice(decision.first).max;
        }

        if(sum >= _inputData->chooser_count())
        {
            continue;
        }

        criticalSets.push_back(s);
    }

    return criticalSets;
}

int SchedulingSolver::slot_order_heuristic_score(map<int, int> const& decisions, int set)
{
    int score = 0;
    for(auto const& decision : decisions)
    {
        if(decision.second != set) continue;
        score += _inputData->choice(decision.first).max;
    }

    return score;
}

vector<int>
SchedulingSolver::calculate_feasible_slots(map<int, int> const& decisions, vector<bool> const& lowPrioritySlot, int choice)
{
    // Feasible slots are all slots for which adding the current choice would not cause the
    // minimal chooser number of this slot to exceed the total chooser count.
    //
    // We then have to filter the feasible slots by all additional constraints.
    //
    // We order the feasible slots by the maximal chooser number as a heuristic to get more
    // balanced schedulings.
    //
    vector<int> normalSlots, lowSlots;
    vector<int> slotScore(_inputData->slot_count(), INT_MIN);

    if (_inputData->choice(choice).isOptional && satisfies_scheduling_constraints(choice, Scheduling::NOT_SCHEDULED, decisions))
    {
        lowSlots.push_back(Scheduling::NOT_SCHEDULED);
    }

    for(int s = 0; s < _inputData->slot_count(); s++)
    {
        int sum = _inputData->choice(choice).min;
        for(auto const& decision : decisions)
        {
            if(decision.second != s) continue;
            sum += _inputData->choice(decision.first).min;
        }

        if(sum > _inputData->chooser_count() || !satisfies_scheduling_constraints(choice, s, decisions))
        {
            continue;
        }

        if (lowPrioritySlot[s])
        {
            lowSlots.push_back(s);
        }
        else
        {
            normalSlots.push_back(s);
            slotScore[s] = slot_order_heuristic_score(decisions, s);
        }
    }

    std::sort(normalSlots.begin(), normalSlots.end(),
              [&](int const& s1, int const& s2) { return slotScore[s1] < slotScore[s2]; });

    return riffle_shuffle(normalSlots, lowSlots);
}

vector<int> SchedulingSolver::get_choice_scramble()
{
    vector<int> choiceScramble(_inputData->choice_count());
    std::iota(choiceScramble.begin(), choiceScramble.end(), 0);
    std::shuffle(choiceScramble.begin(), choiceScramble.end(), Rng::engine());
    std::sort(choiceScramble.begin(), choiceScramble.end(), [&](int const& x, int const& y)
    {
        return _inputData->scheduling_constraints(x).size() > _inputData->scheduling_constraints(y).size();
    });

    return choiceScramble;
}

vector<bool> SchedulingSolver::get_low_priority_slots()
{
    // TODO: Remove this function?
    vector<bool> res(_inputData->slot_count());
    std::fill(res.begin(), res.end(), false);
    return res;
}

vector<vector<int>> SchedulingSolver::convert_decisions(map<int, int> const& decisions)
{
    vector<vector<int>> res(_inputData->slot_count());
    for(auto const& decision : decisions)
    {
        if (decision.second < 0) continue;
        res[decision.second].push_back(decision.first);
    }

    return res;
}

vector<vector<int>> SchedulingSolver::solve_scheduling(vector<CriticalSet> const& criticalSets, datetime timeLimit)
{
    vector<int> choiceScramble = get_choice_scramble();
    vector<bool> lowPrioritySet = get_low_priority_slots();

    map<int, int> decisions;
    stack<vector<int>> backtracking;

    for(int depth = 0; depth < choiceScramble.size();)
    {
        if(time_now() > timeLimit || is_set(_cancellation))
        {
            return {};
        }

        int choice = choiceScramble[depth];

        if(backtracking.size() <= depth)
        {
            int availableMaxPush = calculate_available_max_push(choiceScramble, depth);

            // If there are any impossibilities, the current partial solution is infeasible.
            //
            if(has_impossibilities(decisions, availableMaxPush))
            {
                backtracking.push({});
            }
            else
            {
                // If the partial solution does not satisfy critical set constraints it is infeasible.
                //
                // This is the case when there aren't enough elements in a critical set to cover all sets. For
                // example, for 4 Slots and the critical set {A, B, C, D, E}, a partial solution of the form
                //
                //      Slot 1:    .. A, C, E ..
                //      Slot 2:    .. D ........
                //      Slot 3:    .............
                //      Slot 4:    .............
                //  Not assigned:  .. B ........
                //
                // Would not be feasible, because the critical set can not be covered anymore (we would need at
                // least 2 open choices in the critical set to cover Set 3 and 4).
                //
                if(!satisfies_critical_sets(decisions, criticalSets))
                {
                    backtracking.push({});
                }
                else
                {
                    vector<int> criticalSlots = calculate_critical_slots(decisions, availableMaxPush, choice);

                    if(criticalSlots.size() == 1)
                    {
                        if (satisfies_scheduling_constraints(choice, criticalSlots[0], decisions))
                        {
                            backtracking.push(criticalSlots);
                        }
                        else
                        {
                            backtracking.push({});
                        }
                    }
                    else if(criticalSlots.size() > 1)
                    {
                        backtracking.push({});
                    }
                    else
                    {
                        vector<int> feasibleSets = calculate_feasible_slots(decisions, lowPrioritySet, choice);
                        backtracking.push(feasibleSets);
                    }
                }
            }
        }

        if(backtracking.top().empty())
        {
            if(depth == 0)
            {
                // no solution
                //
                return {};
            }

            // backtrack
            //
            backtracking.pop();
            decisions.erase(choiceScramble[depth - 1]);
            depth--;
            continue;
        }

        int nextSet = backtracking.top().front();
        auto& top = backtracking.top();
        top.erase(std::remove(top.begin(), top.end(), nextSet), top.end());
        decisions[choice] = nextSet;
        depth++;
    }

    return convert_decisions(decisions);
}

SchedulingSolver::SchedulingSolver(const_ptr<InputData> inputData,
                                   const_ptr<CriticalSetAnalysis> csAnalysis,
                                   const_ptr<Options> options,
                                   cancel_token cancellation)
        : _inputData(std::move(inputData)),
          _csAnalysis(std::move(csAnalysis)),
          _currentSolution(new Scheduling(_inputData)),
          _hasSolution(false),
          _options(std::move(options)),
          _cancellation(std::move(cancellation))
{
}

bool SchedulingSolver::next_scheduling()
{
    int preferenceLimit = Rng::next(0, PREF_RELAXATION) == 0
                          ? _inputData->max_preference()
                          : _csAnalysis->preference_bound();

    vector<vector<int>> slots;

    while(slots.empty())
    {
        vector<CriticalSet> csSets = _csAnalysis->for_preference(preferenceLimit);

        datetime timeLimit =  preferenceLimit == _inputData->max_preference()
                              ? time_never()
                              : time_now() + seconds(_options->critical_set_timeout_seconds());

        slots = solve_scheduling(csSets, timeLimit);

        if(slots.empty())
        {
            if(preferenceLimit == _inputData->max_preference())
            {
                _hasSolution = false;
                _currentSolution = nullptr;
                return false;
            }

            preferenceLimit = _inputData->preference_after(preferenceLimit);
        }
    }

    vector<int> scheduling(_inputData->choice_count(), -1);
    for(int s = 0; s < slots.size(); s++)
    {
        for(int w = 0; w < slots[s].size(); w++)
        {
            scheduling[slots[s][w]] = s;
        }
    }

    _hasSolution = true;
    _currentSolution = std::make_shared<Scheduling>(_inputData, scheduling);

    return true;
}

File: src/scheduling_solver.rs
use std::cmp::{Reverse, max, min};
use std::time::{Duration, SystemTime};

use crate::shotgun_solver::ProgressReporter;
use crate::status;
use crate::util::{riffle_shuffle, time_never, time_now};
use crate::{ConstraintType, CriticalSet, Options, PreparedProblem, Rng, Scheduling};

/// Enumerates feasible schedulings for a given input.
#[derive(Debug)]
pub struct SchedulingSolver<'a> {
    problem: &'a PreparedProblem,
    options: &'a Options,
    current_solution: Option<Scheduling>,
    has_solution: bool,
    progress_reporter: ProgressReporter,
    rng: Rng,
    pub(crate) current_depth: usize,
    pub(crate) max_depth_reached: usize,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum SchedulingDecision {
    Undecided,
    Omitted,
    Slot(usize),
}

#[derive(Debug, Clone)]
struct SchedulingSearchState {
    decisions: Vec<SchedulingDecision>,
    slot_min: Vec<u32>,
    slot_max: Vec<u32>,
    slot_choice_count: Vec<u32>,
}

impl SchedulingDecision {
    fn slot(self) -> Option<usize> {
        match self {
            Self::Undecided | Self::Omitted => None,
            Self::Slot(slot) => Some(slot),
        }
    }
}

impl SchedulingSearchState {
    fn new(choice_count: usize, slot_count: usize) -> Self {
        Self {
            decisions: vec![SchedulingDecision::Undecided; choice_count],
            slot_min: vec![0; slot_count],
            slot_max: vec![0; slot_count],
            slot_choice_count: vec![0; slot_count],
        }
    }
}

impl<'a> SchedulingSolver<'a> {
    pub(crate) const PREF_RELAXATION: i32 = 10;

    /// Creates a new scheduling solver for the prepared problem.
    #[must_use]
    pub fn new(problem: &'a PreparedProblem, options: &'a Options) -> Self {
        Self::new_with_progress(
            problem,
            options,
            ProgressReporter::dummy(),
            Rng::from_global_seed(),
        )
    }

    pub(crate) fn new_with_rng(
        problem: &'a PreparedProblem,
        options: &'a Options,
        progress_reporter: ProgressReporter,
        rng: Rng,
    ) -> Self {
        Self::new_with_progress(problem, options, progress_reporter, rng)
    }

    fn new_with_progress(
        problem: &'a PreparedProblem,
        options: &'a Options,
        progress_reporter: ProgressReporter,
        rng: Rng,
    ) -> Self {
        Self {
            problem,
            options,
            current_solution: Some(Scheduling::new()),
            has_solution: false,
            progress_reporter,
            rng,
            current_depth: 0,
            max_depth_reached: 0,
        }
    }

    /// Advances the search to the next feasible scheduling.
    ///
    /// # Panics
    ///
    /// Panics if `critical_set_timeout_seconds` cannot be converted to `u64`.
    pub fn next_scheduling(&mut self, deadline: Option<SystemTime>) -> bool {
        let input_data = &self.problem.input_data;
        self.current_depth = 0;
        self.max_depth_reached = 0;
        let mut preference_limit = if self.rng.next_in_range(0, Self::PREF_RELAXATION) == 0 {
            input_data.max_preference
        } else {
            self.problem.critical_set_analysis.preference_bound()
        };
        status::trace(&format!(
            "Starting scheduling search with preference limit {preference_limit}."
        ));

        let slots = loop {
            let cs_sets = self
                .problem
                .critical_set_analysis
                .for_preference(preference_limit);
            let time_limit = if preference_limit == input_data.max_preference {
                time_never()
            } else {
                time_now()
                    + Duration::from_secs(
                        u64::try_from(self.options.critical_set_timeout_seconds)
                            .expect("critical set timeout must be non-negative"),
                    )
            };
            let time_limit = deadline.map_or(time_limit, |deadline| min(time_limit, deadline));

            let slots = self.solve_scheduling(cs_sets, time_limit);
            if !slots.is_empty() {
                status::trace(&format!(
                    "Scheduling search succeeded with preference limit {preference_limit}."
                ));
                break slots;
            }

            if preference_limit == input_data.max_preference {
                status::debug(
                    "Scheduling search exhausted all preference limits without a solution.",
                );
                self.has_solution = false;
                self.current_solution = None;
                return false;
            }

            status::trace(&format!(
                "No scheduling found at preference limit {preference_limit}; relaxing limit."
            ));
            preference_limit = input_data.preference_after(preference_limit);
        };

        let mut scheduling = vec![None; input_data.choices.len()];
        for (slot_index, choices) in slots.iter().enumerate() {
            for &choice in choices {
                scheduling[choice] = Some(slot_index);
            }
        }

        self.has_solution = true;
        self.current_depth = input_data.choices.len();
        self.publish_depth_progress();
        self.current_solution = Some(Scheduling::with_data(input_data, scheduling));
        true
    }

    /// Returns the most recently generated scheduling, if any.
    #[must_use]
    pub fn scheduling(&self) -> Option<Scheduling> {
        self.current_solution.clone()
    }

    fn satisfies_critical_sets(
        &self,
        decisions: &[SchedulingDecision],
        critical_sets: &[CriticalSet],
    ) -> bool {
        let slot_count = self.problem.input_data.slots.len();
        for set in critical_sets {
            let mut covered_slots = vec![false; slot_count];
            let mut covered_count = 0_usize;
            let mut missing = 0_usize;

            for &element in &set.data {
                match decisions[element] {
                    SchedulingDecision::Undecided => missing += 1,
                    SchedulingDecision::Slot(slot) => {
                        if !covered_slots[slot] {
                            covered_slots[slot] = true;
                            covered_count += 1;
                        }
                    }
                    SchedulingDecision::Omitted => {}
                }
            }

            if covered_count + missing < slot_count {
                return false;
            }
        }

        true
    }

    fn satisfies_scheduling_constraints(
        &self,
        choice: usize,
        slot: Option<usize>,
        state: &SchedulingSearchState,
    ) -> bool {
        let input_data = &self.problem.input_data;
        for constraint in input_data.scheduling_constraints_for(choice) {
            match constraint.kind {
                ConstraintType::ChoiceIsInSlot => {
                    if slot != Some(constraint.slot()) {
                        return false;
                    }
                }
                ConstraintType::ChoiceIsNotInSlot => {
                    if slot == Some(constraint.slot()) {
                        return false;
                    }
                }
                ConstraintType::ChoicesAreInSameSlot => {
                    let other = if constraint.left == choice {
                        constraint.other_choice()
                    } else {
                        constraint.left
                    };
                    let other_slot = state.decisions[other].slot();
                    if state.decisions[other] != SchedulingDecision::Undecided && other_slot != slot
                    {
                        return false;
                    }
                }
                ConstraintType::ChoicesAreNotInSameSlot => {
                    let other = if constraint.left == choice {
                        constraint.other_choice()
                    } else {
                        constraint.left
                    };
                    let other_slot = state.decisions[other].slot();
                    if state.decisions[other] != SchedulingDecision::Undecided
                        && other_slot == slot
                        && slot.is_some()
                    {
                        return false;
                    }
                }
                ConstraintType::ChoicesHaveOffset => {
                    let other = if constraint.left == choice {
                        constraint.other_choice()
                    } else {
                        constraint.left
                    };
                    let offset = if other == constraint.left {
                        -constraint.offset()
                    } else {
                        constraint.offset()
                    };
                    if state.decisions[other] != SchedulingDecision::Undecided {
                        let other_slot = state.decisions[other].slot();
                        if other_slot.is_some() != slot.is_some() {
                            return false;
                        }
                        let (Some(other_slot), Some(slot)) = (other_slot, slot) else {
                            continue;
                        };
                        if i32::try_from(other_slot).expect("slot index must fit in i32")
                            - i32::try_from(slot).expect("slot index must fit in i32")
                            != offset
                        {
                            return false;
                        }

                        let min_slot = max(0, -offset);
                        let max_slot = min(
                            i32::try_from(input_data.slots.len())
                                .expect("slot count must fit in i32"),
                            i32::try_from(input_data.slots.len())
                                .expect("slot count must fit in i32")
                                - offset,
                        );
                        let slot = i32::try_from(slot).expect("slot index must fit in i32");
                        if slot < min_slot || slot >= max_slot {
                            return false;
                        }
                    }
                }
                ConstraintType::SlotHasLimitedSize => {
                    if slot != Some(constraint.left) {
                        continue;
                    }

                    if matches!(
                        constraint.slot_size_limit_op(),
                        crate::SlotSizeLimitOp::Gt
                            | crate::SlotSizeLimitOp::Geq
                            | crate::SlotSizeLimitOp::Neq
                    ) {
                        continue;
                    }

                    let mut limit = i32::try_from(constraint.limit())
                        .expect("limit must fit in i32")
                        - i32::from(matches!(
                            constraint.slot_size_limit_op(),
                            crate::SlotSizeLimitOp::Lt
                        ));
                    limit -= i32::try_from(slot.map_or(0, |slot| state.slot_choice_count[slot]))
                        .expect("slot choice count must fit in i32");
                    if limit < 1 {
                        return false;
                    }
                }
                _ => panic!("Unknown scheduling type {:?}.", constraint.kind),
            }
        }

        if state
            .decisions
            .iter()
            .filter(|&&decision| decision != SchedulingDecision::Undecided)
            .count()
            + 1
            == input_data.choices.len()
        {
            self.check_slot_size_constraints(slot, &state.slot_choice_count)
        } else {
            true
        }
    }

    fn check_slot_size_constraints(&self, slot: Option<usize>, slot_choice_count: &[u32]) -> bool {
        let input_data = &self.problem.input_data;

        for constraint in &input_data.scheduling_constraints {
            if constraint.kind != ConstraintType::SlotHasLimitedSize {
                continue;
            }
            let count =
                slot_choice_count[constraint.left] + u32::from(slot == Some(constraint.left));
            let valid = match constraint.slot_size_limit_op() {
                crate::SlotSizeLimitOp::Eq => count == constraint.limit(),
                crate::SlotSizeLimitOp::Neq => count != constraint.limit(),
                crate::SlotSizeLimitOp::Lt => count < constraint.limit(),
                crate::SlotSizeLimitOp::Leq => count <= constraint.limit(),
                crate::SlotSizeLimitOp::Gt => count > constraint.limit(),
                crate::SlotSizeLimitOp::Geq => count >= constraint.limit(),
            };
            if !valid {
                return false;
            }
        }

        true
    }

    fn has_impossibilities(&self, state: &SchedulingSearchState, available_max_push: u32) -> bool {
        let input_data = &self.problem.input_data;
        for slot in 0..input_data.slots.len() {
            let sum = available_max_push + state.slot_max[slot];
            if sum
                < u32::try_from(input_data.choosers.len()).expect("chooser count must fit in u32")
            {
                return true;
            }
        }
        false
    }

    fn calculate_critical_slots(
        &self,
        state: &SchedulingSearchState,
        available_max_push: u32,
        choice: usize,
    ) -> Vec<Option<usize>> {
        let input_data = &self.problem.input_data;
        let mut critical_slots = Vec::new();
        for slot in 0..input_data.slots.len() {
            let sum = available_max_push - input_data.choices[choice].max + state.slot_max[slot];
            if sum
                < u32::try_from(input_data.choosers.len()).expect("chooser count must fit in u32")
            {
                critical_slots.push(Some(slot));
            }
        }
        critical_slots
    }

    fn slot_order_heuristic_score(
        &self,
        state: &SchedulingSearchState,
        slot: Option<usize>,
    ) -> u32 {
        slot.map_or(0, |slot| state.slot_max[slot])
    }

    fn calculate_feasible_slots(
        &mut self,
        state: &SchedulingSearchState,
        low_priority_slot: &[bool],
        choice: usize,
    ) -> Vec<Option<usize>> {
        let input_data = &self.problem.input_data;
        let mut normal_slots = Vec::new();
        let mut low_slots = Vec::new();

        if input_data.choices[choice].is_optional
            && self.satisfies_scheduling_constraints(choice, None, state)
        {
            low_slots.push(None);
        }

        for (slot, &is_low_priority) in low_priority_slot
            .iter()
            .enumerate()
            .take(input_data.slots.len())
        {
            let sum = input_data.choices[choice].min + state.slot_min[slot];

            if sum
                > u32::try_from(input_data.choosers.len()).expect("chooser count must fit in u32")
                || !self.satisfies_scheduling_constraints(choice, Some(slot), state)
            {
                continue;
            }

            if is_low_priority {
                low_slots.push(Some(slot));
            } else {
                normal_slots.push(Some(slot));
            }
        }

        normal_slots.sort_by_key(|&slot| (self.slot_order_heuristic_score(state, slot), slot));
        riffle_shuffle(normal_slots, low_slots, &mut self.rng)
    }

    fn get_choice_scramble(&mut self) -> Vec<usize> {
        let input_data = &self.problem.input_data;
        let mut scramble = (0..input_data.choices.len()).collect::<Vec<_>>();
        self.rng.shuffle(&mut scramble);
        scramble
            .sort_by_key(|&choice| Reverse(input_data.scheduling_constraints_for(choice).len()));
        scramble
    }

    fn get_low_priority_slots(&self) -> Vec<bool> {
        vec![false; self.problem.input_data.slots.len()]
    }

    fn convert_decisions(&self, decisions: &[SchedulingDecision]) -> Vec<Vec<usize>> {
        let mut result = vec![Vec::new(); self.problem.input_data.slots.len()];
        for (choice, &decision) in decisions.iter().enumerate() {
            if let SchedulingDecision::Slot(slot) = decision {
                result[slot].push(choice);
            }
        }
        result
    }

    fn apply_decision(
        &self,
        state: &mut SchedulingSearchState,
        choice: usize,
        slot: Option<usize>,
    ) {
        let previous = state.decisions[choice];
        if let SchedulingDecision::Slot(previous_slot) = previous {
            let data = &self.problem.input_data.choices[choice];
            state.slot_min[previous_slot] -= data.min;
            state.slot_max[previous_slot] -= data.max;
            state.slot_choice_count[previous_slot] -= 1;
        }

        state.decisions[choice] = match slot {
            Some(slot) => {
                let data = &self.problem.input_data.choices[choice];
                state.slot_min[slot] += data.min;
                state.slot_max[slot] += data.max;
                state.slot_choice_count[slot] += 1;
                SchedulingDecision::Slot(slot)
            }
            None => SchedulingDecision::Omitted,
        };
    }

    fn clear_decision(&self, state: &mut SchedulingSearchState, choice: usize) {
        if let SchedulingDecision::Slot(slot) = state.decisions[choice] {
            let data = &self.problem.input_data.choices[choice];
            state.slot_min[slot] -= data.min;
            state.slot_max[slot] -= data.max;
            state.slot_choice_count[slot] -= 1;
        }
        state.decisions[choice] = SchedulingDecision::Undecided;
    }

    fn solve_scheduling(
        &mut self,
        critical_sets: &[CriticalSet],
        time_limit: SystemTime,
    ) -> Vec<Vec<usize>> {
        let input_data = &self.problem.input_data;
        let choice_scramble = self.get_choice_scramble();
        let low_priority_slots = self.get_low_priority_slots();
        let mut state =
            SchedulingSearchState::new(input_data.choices.len(), input_data.slots.len());
        let mut backtracking = Vec::<Vec<Option<usize>>>::new();
        let mut depth = 0_usize;
        let mut available_max_suffix = vec![0_u32; choice_scramble.len() + 1];
        for depth_index in (0..choice_scramble.len()).rev() {
            available_max_suffix[depth_index] = available_max_suffix[depth_index + 1]
                + input_data.choices[choice_scramble[depth_index]].max;
        }

        while depth < choice_scramble.len() {
            self.current_depth = depth;
            self.max_depth_reached = self.max_depth_reached.max(depth);
            self.publish_depth_progress();
            if time_now() > time_limit {
                status::debug("Scheduling search hit its current time limit.");
                return Vec::new();
            }

            let choice = choice_scramble[depth];
            if backtracking.len() <= depth {
                let available_max_push = available_max_suffix[depth];
                let candidates = if self.has_impossibilities(&state, available_max_push)
                    || !self.satisfies_critical_sets(&state.decisions, critical_sets)
                {
                    Vec::new()
                } else {
                    let critical_slots =
                        self.calculate_critical_slots(&state, available_max_push, choice);
                    match critical_slots.len().cmp(&1) {
                        std::cmp::Ordering::Equal => {
                            if self.satisfies_scheduling_constraints(
                                choice,
                                critical_slots[0],
                                &state,
                            ) {
                                critical_slots
                            } else {
                                Vec::new()
                            }
                        }
                        std::cmp::Ordering::Greater => Vec::new(),
                        std::cmp::Ordering::Less => {
                            self.calculate_feasible_slots(&state, &low_priority_slots, choice)
                        }
                    }
                };
                backtracking.push(candidates.into_iter().rev().collect());
            }

            if backtracking[depth].is_empty() {
                if depth == 0 {
                    return Vec::new();
                }

                backtracking.pop();
                self.clear_decision(&mut state, choice_scramble[depth - 1]);
                depth -= 1;
                self.current_depth = depth;
                self.publish_depth_progress();
                continue;
            }

            let next_slot = backtracking[depth]
                .pop()
                .expect("backtracking bucket should contain a candidate");
            self.apply_decision(&mut state, choice, next_slot);
            depth += 1;
            self.current_depth = depth;
            self.max_depth_reached = self.max_depth_reached.max(depth);
            self.publish_depth_progress();
        }

        let result = self.convert_decisions(&state.decisions);
        let mut scheduling = vec![None; input_data.choices.len()];
        for (slot_index, choices) in result.iter().enumerate() {
            for &choice in choices {
                scheduling[choice] = Some(slot_index);
            }
        }
        if !Scheduling::with_data(input_data, scheduling).is_feasible(input_data) {
            return Vec::new();
        }
        result
    }

    fn publish_depth_progress(&self) {
        self.progress_reporter
            .publish_depth(u16::try_from(self.current_depth).map_or(f32::INFINITY, f32::from));
    }
}
